File: JDK11/opentraceability/src/main/java/opentraceability/models/events/transformationevent`1.java
package opentraceability.models.events;

import opentraceability.interfaces.IEventKDE;
import opentraceability.interfaces.ITransformationEvent;
import opentraceability.utility.attributes.*;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;

public class TransformationEvent<T> extends EventBase implements ITransformationEvent {

    @OpenTraceabilityProductsAttribute(value = "inputQuantityList", version = EPCISVersion.V2, type = EventProductType.Input, order = 8, listType = OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute(value = "inputEPCList", version = EPCISVersion.V2, type = EventProductType.Input, order = 7, listType = OpenTraceabilityProductsListType.EPCList)
    public List<EventProduct> inputs = new ArrayList<>();

    public List<EventProduct> outputs = new ArrayList<>();

    public EventAction action = null;

    @OpenTraceabilityAttribute(namespace = "", name = "transformationID", order = 11)
    public String transformationID = null;

    @OpenTraceabilityAttribute(namespace = "", name = "bizStep", order = 12)
    public URI businessStep = null;

    @OpenTraceabilityAttribute(namespace = "", name = "disposition", order = 13)
    public URI disposition = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(namespace = "", name = "readPoint", order = 14)
    public EventReadPoint readPoint = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(namespace = "", name = "bizLocation", order = 15)
    public EventLocation location = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute(name = "bizTransaction")
    @OpenTraceabilityAttribute(namespace = "", name = "bizTransactionList", order = 16)
    public List<EventBusinessTransaction> bizTransactionList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute(name = "source")
    @OpenTraceabilityAttribute(namespace = "", name = "sourceList", order = 17)
    public List<EventSource> sourceList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute(name = "destination")
    @OpenTraceabilityAttribute(namespace = "", name = "destinationList", order = 18)
    public List<EventDestination> destinationList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute(name = "sensorElement")
    @OpenTraceabilityAttribute(namespace = "", name = "sensorElementList", order = 19)
    public List<SensorElement> sensorElementList = new ArrayList<>();

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(namespace = "", name = "persistentDisposition", order = 20)
    public PersistentDisposition persistentDisposition = null;

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute(namespace = "", name = "ilmd", order = 21)
    public T ilmd = null;

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute(namespace = "", name = "type", order = 0)
    public EventType eventType = EventType.TransformationEvent;

    public List<EventProduct> products = new ArrayList<>();
  

    public void addProduct(EventProduct product) throws Exception {
        switch (product.Type) {
            case Output:
                outputs.add(product);
                break;
            case Input:
                inputs.add(product);
                break;
            default:
                throw new Exception("Transformation event only supports inputs and outputs.");
        }
    }

    @Override
    public <T extends IEventKDE> T getKDE(Class<T> clazz, String ns, String name) {
        return super.getKDE(clazz, ns, name);
    }

    @Override
    public <T extends IEventKDE> T getKDE(Class<T> clazz) {
        return super.getKDE(clazz);
    }

    @Override
    public EventILMD grabILMD() {
        return (EventILMD) ilmd;
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/models/events/transformationevent`1.kt
package opentraceability.models.events

import opentraceability.interfaces.IEventKDE
import opentraceability.interfaces.ITransformationEvent
import opentraceability.utility.attributes.*
import java.util.*
import java.net.URI

class TransformationEvent<T> : EventBase(), ITransformationEvent {

    @OpenTraceabilityProductsAttribute("inputQuantityList", EPCISVersion.V2, EventProductType.Input, 8, OpenTraceabilityProductsListType.QuantityList)
    @OpenTraceabilityProductsAttribute("inputEPCList", EPCISVersion.V2, EventProductType.Input, 7, OpenTraceabilityProductsListType.EPCList)
    var inputs: MutableList<EventProduct> = mutableListOf()

    var outputs: MutableList<EventProduct> = mutableListOf()

    override var action: EventAction? = null

    @OpenTraceabilityAttribute("", "transformationID", 11)
    override var transformationID: String? = null

    @OpenTraceabilityAttribute("", "bizStep", 12)
    override var businessStep: URI? = null

    @OpenTraceabilityAttribute("", "disposition", 13)
    override var disposition: URI? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "readPoint", 14)
    override var readPoint: EventReadPoint?= null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "bizLocation", 15)
    override var location: EventLocation? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute( "bizTransaction")
    @OpenTraceabilityAttribute("", "bizTransactionList", 16)
    override var bizTransactionList: MutableList<EventBusinessTransaction> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute( "source")
    @OpenTraceabilityAttribute("", "sourceList", 17)
    override var sourceList: MutableList<EventSource> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute( "destination")
    @OpenTraceabilityAttribute("", "destinationList", 18)
    override var destinationList: MutableList<EventDestination> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityArrayAttribute( "sensorElement")
    @OpenTraceabilityAttribute("", "sensorElementList", 19)
    override var sensorElementList: MutableList<SensorElement> = mutableListOf()

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "persistentDisposition", 20)
    override var persistentDisposition: PersistentDisposition? = null

    @OpenTraceabilityObjectAttribute
    @OpenTraceabilityAttribute("", "ilmd", 21)
    var ilmd: T? = null

    @OpenTraceabilityXmlIgnoreAttribute
    @OpenTraceabilityAttribute("", "type", 0)
    override var eventType: EventType = opentraceability.models.events.EventType.TransformationEvent


    override var products: MutableList<EventProduct> = mutableListOf()
        get() {
            val products = mutableListOf<EventProduct>()
            products.addAll(inputs)
            products.addAll(outputs)
            return products
        }

    override fun addProduct(product: EventProduct) {
        when (product.Type) {
            EventProductType.Output -> outputs.add(product)
            EventProductType.Input -> inputs.add(product)
            else -> throw Exception("Transformation event only supports inputs and outputs.")
        }
    }

    override fun <T: IEventKDE> getKDE(clazz: Class<T>, ns: String, name: String): T? {
        return super.getKDE<T>(clazz, ns, name)
    }

    override fun <T: IEventKDE> getKDE(clazz: Class<T>): T? {
        return super.getKDE<T>(clazz)
    }


    override fun grabILMD(): EventILMD? {
        return ilmd as EventILMD
    }
}
